// @scania/components/dist/collection/components/table/table-body-cell.js
import { Component, h, Host, Listen, Prop, State } from '@stencil/core';
export class TableBodyCell {
  constructor() {
    this.activeSorting = false;
  }
  // Listen to sortColumnData from table-header-element
  updateTextAlign(event) {
    // Nice usage of array deconstruct
    const [keyValue, textDirection] = event.detail;
    if (this.cellKey === keyValue) {
      this.textAlign = textDirection;
    }
  }
  // Listen to sortColumnData from table-header-element
  updateCellActiveState(event) {
    const keyValue = event.detail;
    this.activeSorting = this.cellKey === keyValue;
  }
  render() {
    return (h(Host, { class: {
        'sdds-table__body-cell': true,
        'sdds-table__body-cell--hover': this.activeSorting,
      }, style: { textAlign: this.textAlign } }, this.cellValue));
  }
  static get is() { return "sdds-body-cell"; }
  static get encapsulation() { return "shadow"; }
  static get originalStyleUrls() { return {
    "$": ["table.scss"]
  }; }
  static get styleUrls() { return {
    "$": ["table.css"]
  }; }
  static get properties() { return {
    "cellValue": {
      "type": "any",
      "mutable": false,
      "complexType": {
        "original": "string | number",
        "resolved": "number | string",
        "references": {}
      },
      "required": false,
      "optional": false,
      "docs": {
        "tags": [],
        "text": ""
      },
      "attribute": "cell-value",
      "reflect": true
    },
    "cellKey": {
      "type": "any",
      "mutable": false,
      "complexType": {
        "original": "any",
        "resolved": "any",
        "references": {}
      },
      "required": false,
      "optional": false,
      "docs": {
        "tags": [],
        "text": ""
      },
      "attribute": "cell-key",
      "reflect": true
    }
  }; }
  static get states() { return {
    "textAlign": {},
    "activeSorting": {}
  }; }
  static get listeners() { return [{
      "name": "bodyCellData",
      "method": "updateTextAlign",
      "target": "body",
      "capture": false,
      "passive": false
    }, {
      "name": "headKey",
      "method": "updateCellActiveState",
      "target": "body",
      "capture": false,
      "passive": false
    }]; }
}
