// @scania/components/dist/collection/components/textfield/textfield.js
import { Component, h, State, Prop, Listen, Event, } from '@stencil/core';
export class Textfield {
  constructor() {
    /** Which input type, text, password or similar */
    this.type = 'text';
    /** Label that will be put inside the input */
    this.labelInside = '';
    /** Placeholder text */
    this.placeholder = '';
    /** Value of the input text */
    this.value = '';
    /** Set input in disabled state */
    this.disabled = false;
    /** Size of the input */
    this.size = '';
    /** With setting */
    this.nominwidth = false;
    /** Name property */
    this.name = '';
    /** Autofocus for input */
    this.autofocus = false;
  }
  // Listener if input enters focus state
  handleFocusIn() {
    this.focusInput = true;
  }
  // Listener if input leaves focus state
  handleFocusOut() {
    this.focusInput = false;
  }
  // Data input event in value prop
  handleInput(e) {
    this.value = e.target.value;
  }
  // Change event isn't a composed:true by default in for input
  handleChange(e) {
    this.customChange.emit(e);
  }
  /** Set the input as focus when clicking the whole textfield with suffix/prefix */
  handleFocusClick() {
    this.textInput.focus();
  }
  render() {
    let className = ' sdds-textfield-input';
    if (this.size === 'md') {
      className += className + '-md';
    }
    if (this.size === 'sm') {
      className += className + '-sm';
    }
    return (h("div", { class: `
        ${this.nominwidth ? 'sdds-form-textfield-nomin' : ''}
        ${this.focusInput
        ? 'sdds-form-textfield sdds-textfield-focus'
        : ' sdds-form-textfield'}
        ${this.value.length > 0 ? 'sdds-textfield-data' : ''}
        ${this.labelInside.length > 0
        ? 'sdds-textfield-container-label-inside'
        : ''}
        ${this.disabled ? 'sdds-form-textfield-disabled' : ''}
        ${this.size == 'md' ? 'sdds-form-textfield-md' : ''}
        ${this.size == 'sm' ? 'sdds-form-textfield-sm' : ''}
        ${this.state == 'error' || this.state == 'success'
        ? `sdds-form-textfield-${this.state}`
        : ''}
        ` },
      h("slot", { name: "sdds-label" }),
      h("div", { onClick: () => this.handleFocusClick(), class: "sdds-textfield-container" },
        h("slot", { name: "sdds-prefix" }),
        h("div", { class: "sdds-textfield-input-container" },
          h("input", { ref: (inputEl) => (this.textInput = inputEl), class: className, type: this.type, disabled: this.disabled, placeholder: this.placeholder, value: this.value, autofocus: this.autofocus, maxlength: this.maxlength, name: this.name, onInput: (e) => this.handleInput(e), onChange: (e) => this.handleChange(e) }),
          this.labelInside.length > 0 && (h("label", { class: "sdds-textfield-label-inside" }, this.labelInside))),
        h("div", { class: "sdds-textfield-bar" }),
        h("slot", { name: "sdds-suffix" })),
      h("div", { class: "sdds-textfield-helper" },
        h("slot", { name: "sdds-helper" }),
        this.maxlength > 0 && (h("div", { class: "sdds-textfield-textcounter" },
          this.value.length,
          h("span", { class: "sdds-textfield-textcounter-divider" }, " / "),
          this.maxlength)))));
  }
  static get is() { return "sdds-textfield"; }
  static get encapsulation() { return "shadow"; }
  static get originalStyleUrls() { return {
    "$": ["textfield.scss"]
  }; }
  static get styleUrls() { return {
    "$": ["textfield.css"]
  }; }
  static get properties() { return {
    "type": {
      "type": "string",
      "mutable": false,
      "complexType": {
        "original": "string",
        "resolved": "string",
        "references": {}
      },
      "required": false,
      "optional": false,
      "docs": {
        "tags": [],
        "text": "Which input type, text, password or similar"
      },
      "attribute": "type",
      "reflect": true,
      "defaultValue": "'text'"
    },
    "labelInside": {
      "type": "string",
      "mutable": false,
      "complexType": {
        "original": "string",
        "resolved": "string",
        "references": {}
      },
      "required": false,
      "optional": false,
      "docs": {
        "tags": [],
        "text": "Label that will be put inside the input"
      },
      "attribute": "label-inside",
      "reflect": false,
      "defaultValue": "''"
    },
    "placeholder": {
      "type": "string",
      "mutable": false,
      "complexType": {
        "original": "string",
        "resolved": "string",
        "references": {}
      },
      "required": false,
      "optional": false,
      "docs": {
        "tags": [],
        "text": "Placeholder text"
      },
      "attribute": "placeholder",
      "reflect": false,
      "defaultValue": "''"
    },
    "value": {
      "type": "string",
      "mutable": false,
      "complexType": {
        "original": "string",
        "resolved": "string",
        "references": {}
      },
      "required": false,
      "optional": false,
      "docs": {
        "tags": [],
        "text": "Value of the input text"
      },
      "attribute": "value",
      "reflect": true,
      "defaultValue": "''"
    },
    "disabled": {
      "type": "boolean",
      "mutable": false,
      "complexType": {
        "original": "boolean",
        "resolved": "boolean",
        "references": {}
      },
      "required": false,
      "optional": false,
      "docs": {
        "tags": [],
        "text": "Set input in disabled state"
      },
      "attribute": "disabled",
      "reflect": false,
      "defaultValue": "false"
    },
    "size": {
      "type": "string",
      "mutable": false,
      "complexType": {
        "original": "string",
        "resolved": "string",
        "references": {}
      },
      "required": false,
      "optional": false,
      "docs": {
        "tags": [],
        "text": "Size of the input"
      },
      "attribute": "size",
      "reflect": false,
      "defaultValue": "''"
    },
    "nominwidth": {
      "type": "boolean",
      "mutable": false,
      "complexType": {
        "original": "boolean",
        "resolved": "boolean",
        "references": {}
      },
      "required": false,
      "optional": false,
      "docs": {
        "tags": [],
        "text": "With setting"
      },
      "attribute": "nominwidth",
      "reflect": false,
      "defaultValue": "false"
    },
    "name": {
      "type": "string",
      "mutable": false,
      "complexType": {
        "original": "string",
        "resolved": "string",
        "references": {}
      },
      "required": false,
      "optional": false,
      "docs": {
        "tags": [],
        "text": "Name property"
      },
      "attribute": "name",
      "reflect": false,
      "defaultValue": "''"
    },
    "state": {
      "type": "string",
      "mutable": false,
      "complexType": {
        "original": "string",
        "resolved": "string",
        "references": {}
      },
      "required": false,
      "optional": false,
      "docs": {
        "tags": [],
        "text": "Error state of input"
      },
      "attribute": "state",
      "reflect": false
    },
    "maxlength": {
      "type": "number",
      "mutable": false,
      "complexType": {
        "original": "number",
        "resolved": "number",
        "references": {}
      },
      "required": false,
      "optional": false,
      "docs": {
        "tags": [],
        "text": "Max length of input"
      },
      "attribute": "maxlength",
      "reflect": false
    },
    "autofocus": {
      "type": "boolean",
      "mutable": false,
      "complexType": {
        "original": "boolean",
        "resolved": "boolean",
        "references": {}
      },
      "required": false,
      "optional": false,
      "docs": {
        "tags": [],
        "text": "Autofocus for input"
      },
      "attribute": "autofocus",
      "reflect": false,
      "defaultValue": "false"
    }
  }; }
  static get states() { return {
    "focusInput": {}
  }; }
  static get events() { return [{
      "method": "customChange",
      "name": "customChange",
      "bubbles": true,
      "cancelable": true,
      "composed": true,
      "docs": {
        "tags": [],
        "text": "Change event for the textfield"
      },
      "complexType": {
        "original": "any",
        "resolved": "any",
        "references": {}
      }
    }]; }
  static get listeners() { return [{
      "name": "focus",
      "method": "handleFocusIn",
      "target": undefined,
      "capture": false,
      "passive": false
    }, {
      "name": "focusout",
      "method": "handleFocusOut",
      "target": undefined,
      "capture": false,
      "passive": false
    }]; }
}
